### PR TITLE
[html] Improvements for handling text and comment nodes

### PR DESCRIPTION
**Pull request by adangel**
## Describe the PR

This adds native support for text and comment nodes, so that the XPath selectors `text()` and `comment()` can be used.

This works by relying on the node names "#text" and "#comment".

This means, these nodes appear in the designer as "normal" nodes, but can be selected by `text()` as usual in XPath.

Note: This doesn't affect the XML module, as there we explicitly remove the hash:
https://github.com/pmd/pmd/blob/d3880443f7dbc70718159ce849193c505fa08f52/pmd-xml/src/main/java/net/sourceforge/pmd/lang/xml/ast/XmlNodeWrapper.java#L212-L214

It should not affect any other language, but the solution is to use "#text" and "#comment" as indicators to use a different node type. But "#" is not a common node name for javacc/antlr based languages, so it should have no effect.

## Related issues

- Refs #3941 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)
- [ ] Update release notes

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/saxon/ElementNode.java [hunk 1]
```diff
@@ -80,6 +80,16 @@ public ElementNode(DocumentNode document,
         document.nodeToElementNode.put(node, this);
     }
 
+    private static int determineType(Node node) {
```
Comment (oowekyala) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/saxon/ElementNode.java:83]:
```suggestion
    private static int determineType(Node node) {
        // As of PMD 6.48.0, only the experimental HTML module uses this naming 
        // convention to identify non-element nodes.
        // TODO PMD 7: maybe generalize this to other languages
```

In a general PMD 7 solution, I don't expect we use a naming convention to identify the node kind (we'd probably use an enum).

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/saxon/ElementNode.java [hunk 2]
```diff
@@ -147,10 +157,18 @@ public Value atomize() {
     }
 
     @Override
-    public CharSequence getStringValueCS() {
+    public String getStringValue() {
+        if (determineType(getUnderlyingNode()) == Type.TEXT) {
```
Comment (oowekyala) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/saxon/ElementNode.java:161]: Comment nodes should also take this path: https://www.w3.org/TR/1999/REC-xpath-19991116/#section-Comment-Nodes

Also the kind has already been computed
```suggestion
        if (getNodeKind() == Type.TEXT || getNodeKind() == Type.COMMENT) {
```

File: pmd-html/src/test/java/net/sourceforge/pmd/lang/html/HtmlXPathRuleTest.java [hunk 3]
```diff
@@ -24,25 +20,42 @@ public class HtmlXPathRuleTest {
...
+            + "      <p>Test</p>\n"
+            + "    </template>\n"
             + "</template>";
 
     @Test
     public void selectTextNode() {
         // from https://developer.salesforce.com/docs/component-library/documentation/en/lwc/lwc.js_props_getter
         // "Don’t add spaces around the property, for example, { data } is not valid HTML."
-        String xpath = "//*[local-name() = '#text'][contains(@Text, '{ ')]";
+        String xpath = "//text()[contains(., '{ ')]";
```
Comment (oowekyala) [inline pmd-html/src/test/java/net/sourceforge/pmd/lang/html/HtmlXPathRuleTest.java:38]: cool :) Can we also assert that the previous query doesn't match anything? Ie, saxon does not see elements with name `#text` or `#comment`

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/3ea2f5f018e34d2fa45b84324153432b/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-html/src/test/java/net/sourceforge/pmd/lang/html/HtmlXPathRuleTest.java (hunk 3 above)
Comment (adangel) [reply pmd-html/src/test/java/net/sourceforge/pmd/lang/html/HtmlXPathRuleTest.java:38]: I've added another test case - it works, `#text` doesn't match anything anymore.

**Pull request merged**